[tests] Adjust limits test to dynamic limits

gahaas opened the pull request:
This PR changes the tests in limits.any.js for memory size and table size to validate and compile modules that exceed implementation-defined limits. Only during instantiation and runtime, errors should be thrown.

In https://github.com/WebAssembly/spec/pull/1195 @Ms2ger requested that these tests get adjusted to to changes to the spec.

gahaas commented:
ping

Ms2ger commented:
Trying to get to it this week or next

gahaas commented:
ping

gahaas commented on test/js-api/limits.any.js:
```diff
@@ -167,30 +150,81 @@ testLimit("function returns", 0, kJSEmbeddingMaxFunctionReturns, (builder, count
+  promise_test(t => { return WebAssembly.compile(buffer); },
+               `Async compile ${name} beyond its dynamic limit.`);
+
+  if (test_instantiation) {
```
I tested instantiation only for those values where instantiation should fail, e.g. initial memory size.

Ms2ger replied:
Hmm, should we try testing that instantiation *doesn't* fail for the others, or will those fail anyway because the actual limit will be lower?

gahaas replied:
I added tests now for succeeding instantiation, and also for growing.

Ms2ger commented on test/js-api/limits.any.js:
```diff
@@ -167,30 +150,81 @@ testLimit("function returns", 0, kJSEmbeddingMaxFunctionReturns, (builder, count
+  }
+}
+
+testDynamicLimit(
+    "initial declared memory pages", kJSEmbeddingMaxMemoryPages, true,
+    (builder, count) => { builder.addMemory(count, undefined, false, false); });
```
We should also test the case where a memory is initially within the limit, but then is grown past the limit, including with the `memory.grow` instruction in wasm proper, and with `WebAssembly.Memory.grow()`.

However, note that the spec already requires a `RangeError` if `mem_grow` fails. It doesn't seem well-defined whether `RangeError` or `RuntimeError` wins here.

gahaas replied:
I added tests for growing now. I changed the expectation to RangeError.

Ms2ger commented on test/js-api/limits.any.js:
```diff
@@ -167,30 +150,81 @@ testLimit("function returns", 0, kJSEmbeddingMaxFunctionReturns, (builder, count
+    "maximum imported memory pages", kJSEmbeddingMaxMemoryPages, false,
+    (builder, count) => { builder.addImportedMemory("mod", "mem", 1, count); });
+
+testDynamicLimit(
+    "initial table size", kJSEmbeddingMaxTableSize, true,
+    (builder, count) => { builder.setTableBounds(count, undefined); });
```
Same about growing.

gahaas replied:
Done.

gahaas commented:
I updated the test now to test both failing and succeeding instantiation, and also growing beyond the implementation-defined limits. Sorry that it took so long.

I also changed the values for `kJSEmbeddingMaxFunctionReturns` and `kJSEmbeddingMaxTables`, because with recent additions to the spec, these values are higher now.

gahaas commented:
I just realized that the reference types proposal is not merged yet. I reverted the number of allowed tables to 1, and removed the table.grow test.

gahaas commented:
I removed the memory limits tests now. The implementation-defined memory limit is the same as the memory limit of the core spec. Therefore this limit is already covered by the core spec tests.

gahaas commented:
@Ms2ger ping

gahaas commented:
@Ms2ger ping
